native/src/main/java/io/ballerina/stdlib/kafka/utils/KafkaUtils.java

```diff
@@ -634,18 +639,43 @@ private static void handleHeaderValuesWithArrayType(String key, ArrayList<byte[]
         Type elementType = arrayType.getElementType();
         if (elementType.getTag() == ARRAY_TAG) {
             BArray valueArray = ValueCreator.createArrayValue(arrayType);
+            if (Objects.isNull(list)) {
+                bHeaderMap.put(StringUtils.fromString(key), null);
+                return;
+            }
             for (int i = 0; i < list.size(); i++) {
-                valueArray.add(i, ValueCreator.createArrayValue(list.get(i)));
+                byte[] value = list.get(i);
+                if (Objects.isNull(value)) {
+                    // todo: identify how to do this
+                    valueArray.append(null);
+                    continue;
+                }
+                valueArray.add(i, ValueCreator.createArrayValue(value));
             }
             bHeaderMap.put(StringUtils.fromString(key), valueArray);
         } else if (elementType.getTag() == STRING_TAG) {
```

Comment:
I think we have to consider the element type being `string?`; union type, in order to support this.